File: src/overlays.js
const fs = require('fs-extra');
const path = require('path');
const http = require('http');
const https = require('https');
const events = require('events');
const settings = require('electron-settings');

const Downloader = require('./downloader.js');
const downloader = new Downloader();

const data = require('./data.json');

let mustCancel = false; // cancellation token
let nbOverlays = 0; // number of installed overlays
let total = 0; // total number of processed items
let current = 0; // current item number

module.exports = class Overlays extends events {
    /**
     * Asks the class to cancel the next action
     */
    cancel() {
        mustCancel = true;
    }

    /**
     * Checks that the program can access the specified folder
     * 
     * @param {string} folder The folder to check for access
     * @param {string} checkWrite Whether to check for write access
     * @returns {bool} Whether the check is successful
     */
    checkAccess (folder, checkWrite) {
        try {
            fs.ensureDirSync(folder);
        } catch (err) {
            console.error('The folder %s does not exist or cannot be read, and cannot be created!', folder);
            console.log('Exiting...');
            return false;
        }
        
        if (fs.existsSync(folder)) {
            if (checkWrite) {
                var testFile = path.join(folder, '_test-install.txt');
                try {
                    fs.writeFileSync(testFile, 'test');
                    if (fs.existsSync(testFile)) {
                        fs.unlinkSync(testFile);
                    } else {
                        console.error('Unable to write files into the folder %s!', folder);
                        console.log('Exiting...');
                        return false;
                    }
                } catch (err) {
                    console.error('Unable to write files into the folder %s: %o', folder, err);
                    console.log('Exiting...');
                    return false;
                }
    
                console.log('%s can be written to', folder);
            } else {
                console.log('%s can be read', folder);
            }
        } else {
            console.error('The folder %s does not exist or cannot be read!', folder);
            console.log('Exiting...');
            return false;
        }

        return true;
    }

    /**
     * Fixes the paths in the specified file content
     * 
     * @param {object} base The base paths (ex: { retropie: /etc/, recalbox: /etc/ })
     * @param {string} content The file content to fix
     */
    fixPath(base, content) {
        let fromOs = settings.get('os') === 'retropie' ? 'recalbox' : 'retropie';
        let toOs = settings.get('os');
        return content.replace(new RegExp(base[fromOs], 'gm'), base[toOs]);
    }

    /**
     * Extracts a configuration value from the content
     * 
     * @param {String} content The content to search into
     * @param {String} value The value to search for
     * @returns {Number} The configuration value
     */
    parseValue(content, value) {
        let regex = new RegExp(value + '[\\s]*=[\\s]*"?(\\d+)"?', 'igm');
        let result = regex.exec(content);
        if (result && result.length >= 2) {
            return parseInt(result[1]);
        } else {
            return null;
        }
    }

    /**
     * Changes the specified file content resolution
     * 
     * @param {String} content The file content
     * @param {Number} ratio The ratio to multiply width/height/x/y with
     * @returns {String} The modified content
     */
    changeResolution(content, ratio) {
        let parameters = [
            'custom_viewport_width', 'custom_viewport_height',
            'custom_viewport_x', 'custom_viewport_y',
            'video_fullscreen_x', 'video_fullscreen_y'
        ];
        for (let p of parameters) {
            let val = this.parseValue(content, p);
            if (val) {
                val = Math.round(val * ratio);
                content = content.replace(new RegExp(p + '[\\s]*=[\\s]*"?(\\d+)"?', 'img'), p + ' = ' + val);
            }
        }

        return content;
    }

    /**
     * Lists the files in the specified folder
     * 
     * @param {any} repository The repository
     * @param {any} folder The folder to list the files from
     * @returns {Promise} A promise listing the files in the folder
     */
    listFiles(repository, folder) {
        return new Promise((resolve, reject) => {
            downloader.listFiles(repository, folder, (files) => {
                if (mustCancel) { resolve(); return; }
                
                resolve(files);
            });
        });
    }

    /**
     * Downloads the common items
     * 
     * @param {String} repository The repository
     * @param {Object} common The common elements
     * @param {String} destPath The destination folder
     * @param {Boolean} overwrite Whether to overwrite existing files
     * @param {Object} base The base paths (ex: { retropie: /etc/, recalbox: /etc/ })
     * @param {Number} ratio The ratio to multiply width/height/x/y with
     * @returns {Promise} A promise downloading the items in the common folder
     */
    downloadCommon(repository, common, destPath, overwrite, base, ratio) {
        return new Promise((resolve, reject) => {
            if (mustCancel) { reject(); return; }

            if (typeof common !== 'undefined' && common) {
                total++;
                console.log('Installing common files');
                this.emit('progress.download', total, current++, 'common files');

                // only download if target folder does not exist
                if (overwrite || !fs.existsSync(destPath)) {
                    downloader.downloadFolder(repository, common.src, destPath, overwrite, (content) => {
                        return this.changeResolution(this.fixPath(base, content), ratio);
                    }, resolve);
                } else {
                    // no not overwrite and target folder already exists
                    resolve();
                }
            } else {
                // common does not exist
                resolve();
            }
        });
    }

    /**
     * Gets the list of folders containing the specified file
     * 
     * @param {String} folders The folders to search
     * @param {String} file The file to look for
     */
    getFoldersContaining(folders, file) {
        return new Promise((resolve, reject) => {
            let result = [];
            for (let folder of folders) {
                if (fs.existsSync(path.join(folder, file))) {
                    result.push(folder);
                }
            }

            resolve(result);
        });
    }

    /**
     * Downloads the specified rom config
     * 
     * @param {String} repository The repository
     * @param {String} sourcePath The source path to the rom config to download
     * @param {String} destPath The path to save the rom configs to, if any
     * @param {Array} folders The list of folders where to download the config to
     * @param {Boolean} overwrite Whether to overwrite existing files
     * @param {Object} base The base paths
     * @param {Number} ratio The ratio to multiply width/height/x/y with
     * @returns {Promise} A promise that downloads the rom config
     */
    downloadRomConfig(repository, sourcePath, destPath, folders, overwrite, base, ratio) {
        return new Promise((resolveConfig) => {
            let cfg = path.basename(sourcePath);
            let zip = cfg.replace('.cfg', '');

            let romConfigContent = '';

            this.getFoldersContaining(folders, zip)
            .then((romFolders) => {
                let foldersPromises = folders.reduce((promisechain, folder, index) => {
                    return promisechain.then(() => new Promise((resolve) => {
                        let localromcfgPath = destPath;
                        if (localromcfgPath) {
                            // save in config: ex: config/mame, config/fba
                            localromcfgPath = path.join(localromcfgPath, path.basename(folder));
                        } else {
                            // save in rom folder
                            localromcfgPath = folder;
                        }

                        let localromcfg = path.join(localromcfgPath, cfg);
                        if (fs.existsSync(path.join(folder, zip))) {
                            // corresponding zip file exists
                            console.log('Installing overlay for %s in %s', zip, folder);

                            downloader.downloadFile(repository, sourcePath, (romcfgContent) => {
                                if (mustCancel) { resolve(); return; }

                                romcfgContent = this.changeResolution(this.fixPath(base, romcfgContent), ratio);
                                romConfigContent = romcfgContent;
                                if (!overwrite && fs.existsSync(localromcfg)) {
                                    resolve();
                                } else {
                                    // download rom cfg
                                    fs.ensureDirSync(path.dirname(localromcfg));
                                    fs.writeFile(localromcfg, romcfgContent, (err) => {
                                        if (err && err.code !== 'EEXIST') throw err;
                                        if (mustCancel) { resolve(); return; }

                                        resolve();
                                    });
                                }
                            });
                        } else {
                            // the corresponding zip file does not exist
                            resolve();
                        }
                    }));
                }, Promise.resolve());

                // execute promises
                foldersPromises
                .then(() => {
                    resolveConfig(romConfigContent);
                });
            });
        });
    }

    /**
     * Downloads the specified overlay config file to the specified location
     * 
     * @param {String} repository The repository
     * @param {String} sourcePath The source path to the overlay config to download
     * @param {String} destPath The path to the file to write to
     * @param {Boolean} overwrite Whether to overwrite existing files
     * @param {Object} base The base paths
     * @returns {Promise} A promise that downloads the overlay config
     */
    downloadOverlay(repository, sourcePath, destPath, overwrite, base) {
        return new Promise((resolve, reject) => {
            downloader.downloadFile(repository, sourcePath, (packOverlayFileContent) => {
                if (mustCancel) { resolve(); return; }

                if (!overwrite && fs.existsSync(destPath)) {
                    // overlay cfg already exists
                    resolve(packOverlayFileContent);
                } else {
                    // download overlay cfg
                    packOverlayFileContent = this.fixPath(base, packOverlayFileContent);
                    fs.ensureDirSync(path.dirname(destPath));
                    fs.writeFile(destPath, packOverlayFileContent, (err) => {
                        if (err) throw err;
                        if (mustCancel) { resolve(); return; }

                        resolve(packOverlayFileContent);
                    });
                }
            });
        });
    }

    /**
     * Downloads the specified image file to the specified location
     * 
     * @param {String} repository The repository
     * @param {String} sourcePath The source path to the overlay image to download
     * @param {String} destPath The path to the file to write to
     * @param {Boolean} overwrite Whether to overwrite existing files
     * @returns {Promise} A promise that downloads the overlay image
     */
    downloadOverlayImage(repository, sourcePath, destPath, overwrite) {
        return new Promise((resolve, reject) => {
            if (!overwrite && fs.existsSync(destPath)) {
                // image already exists
                resolve();
            } else {
                // download image
                downloader.downloadFile(repository, sourcePath, (imageContent) => {
                    if (mustCancel) { resolve(); return; }

                    fs.writeFile(destPath, imageContent, (err) => {
                        if (err) throw err;
                        nbOverlays++;
                        resolve();
                    });
                });
            }
        });
    }

    /**
     * Downloads and installs an overlay pack
     * 
     * @param {Array} romFolders The rom folders to install the overlays for
     * @param {String} configShare The path to the config share
     * @param {Object} packInfos The chosen pack informations
     * @param {Boolean} overwrite Whether to overwrite existing files
     * @param {Number} ratio The ratio to compute the new overlay resolution
     */
    downloadPack (romFolders, configShare, packInfos, overwrite, ratio) {
        mustCancel = false;
        nbOverlays = 0;

        let repository = packInfos.repository,
            roms = packInfos.roms,
            overlays = packInfos.overlays,
            common = packInfos.common,
            base = packInfos.base;

        const os = settings.get('os');

        // check if the destination of rom cfg is the rom folder
        const romCfgFolder = packInfos.roms.dest[os] === 'roms'
            ? null // save rom cfg directly into rom folder(s)
            : path.join(configShare, packInfos.roms.dest[os]); // save rom cfg in config folder

        this.emit('start.download');
        this.emit('progress.download', 100, 1, 'files list');

        this.downloadCommon(repository, common, common ? path.join(configShare, common.dest[os]) : '', overwrite, base, ratio)
        .then(() => {
            return this.listFiles(repository, roms.src);
        })
        .then((romConfigs) => {
            total = romConfigs.length;
            current = 1;

            // loop on each config file
            let requests = romConfigs.reduce((promisechain, romcfg, index) => {
                return promisechain.then(() => new Promise((resolve, reject) => {
                    if (mustCancel) { resolve(); return; }

                    current++;

                    // only process config files
                    if (romcfg.type !== 'file' || !romcfg.name.endsWith('.zip.cfg')) { 
                        resolve();
                        return;
                    }

                    this.emit('progress.download', total, current, romcfg.name.replace('.zip.cfg', ''));
        
                    // download rom config
                    this.downloadRomConfig(repository, romcfg.path, romCfgFolder, romFolders, overwrite, base, ratio)
                    .then((romcfgContent) => {
                        if (!romcfgContent || romcfgContent === '') { return Promise.resolve(''); }

                        // parse rom cfg to get overlay cfg
                        let overlayFile = /input_overlay[\s]*=[\s]*"?(.*\.cfg)"?/igm.exec(romcfgContent)[1]; // extract overlay path
                        overlayFile = overlayFile.substring(overlayFile.lastIndexOf('/')); // just the file name
                        let packOverlayFile = path.join(overlays.src, overlayFile); // concatenate with pack path                          
                        let localoverlaycfg = path.join(configShare, overlays.dest[os], overlayFile);

                        // download overlay config
                        return this.downloadOverlay(repository, packOverlayFile, localoverlaycfg, overwrite, base);
                    })
                    .then((overlayCfgContent) => {
                        if (!overlayCfgContent || overlayCfgContent === '') { return Promise.resolve(); }

                        // parse overlay cfg to get overlay image
                        let packOverlayImage = /overlay0_overlay[\s]*=[\s]*"?(.*\.png)"?/igm.exec(overlayCfgContent)[1];
                        let packOverlayImageFile = path.join(overlays.src, packOverlayImage);
                        let localoverlayimg = path.join(configShare, overlays.dest[os], packOverlayImage);

                        // download overlay image
                        return this.downloadOverlayImage(repository, packOverlayImageFile, localoverlayimg, overwrite);
                    })
                    .then(() => {
                        resolve();
                    });
                }));
            }, Promise.resolve());

            return requests;
        })
        .then(() => {
            this.emit('log', 'Installed ' + nbOverlays + ' overlays');
            this.emit('end.download', mustCancel);
        });
    }
};

File: ArcadeManager/Services/Overlays.cs
using ArcadeManager.Models;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace ArcadeManager.Services {

	public class Overlays : IOverlays {
		private readonly IDownloader downloaderService;

		/// <summary>
		/// Initializes a new instance of the <see cref="Overlays"/> class.
		/// </summary>
		/// <param name="downloaderService">The downloader service.</param>
		public Overlays(IDownloader downloaderService) {
			this.downloaderService = downloaderService;
		}

		/// <summary>
		/// Downloads an overlay pack
		/// </summary>
		/// <param name="data">The parameters</param>
		/// <param name="messageHandler">The message handler.</param>
		/// <exception cref="FileNotFoundException">Unable to parse rom config {game} to find overlay (input_overlay)
		/// or
		/// Unable to parse overlay config {game} to find image (overlay0_overlay)</exception>
		public async Task Download(Actions.OverlaysAction data, IMessageHandler messageHandler) {
			messageHandler.Init("Download overlay pack");

			try {
				var os = ArcadeManagerEnvironment.SettingsOs;
				var pack = ArcadeManagerEnvironment.AppData.Overlays.Where(o => o.Name == data.pack).First();

				// check if the destination of rom cfg is the rom folder
				var romCfgFolder = pack.Roms.Dest[os] == "roms"
					? null // save rom cfg directly into rom folder(s)
					: Path.Join(data.configFolder, pack.Roms.Dest[os]); // save rom cfg in config folder

				// list the available rom configs
				messageHandler.Progress("list of files to download", 1, 100);
				var romConfigs = await downloaderService.ListFiles(pack.Repository, pack.Roms.Src);

				// download common files
				messageHandler.Progress("common files", 1, 100);
				if (pack.Common != null && !string.IsNullOrWhiteSpace(pack.Common.Src)) {
					await DownloadCommon(pack, Path.Join(data.configFolder, pack.Common.Dest[os]), data.overwrite, data.ratio, messageHandler, 100, 1);
				}

				if (messageHandler.MustCancel) { throw new Exception("Operation cancelled"); }

				// check that thvere is a matching game in any of the roms folders
				messageHandler.Progress("games list to process", 1, 100);
				var processedOverlays = new List<string>();
				var romsToProcess = GetRomsToProcess(data.romFolders, romConfigs.Tree);

				var total = romConfigs.Tree.Count;
				var current = 0;
				var installed = 0;

				foreach (var r in romsToProcess.OrderBy(r => r.Game)) {
					if (messageHandler.MustCancel) { throw new Exception("Operation cancelled"); }

					current++;

					var game = r.Game;

					messageHandler.Progress($"{game}: download overlay (rom config)", total, current);

					// download the rom config and extract the overlay file name
					var romConfigContent = string.Empty;
					foreach (var romFolder in r.TargetFolder) {
						if (messageHandler.MustCancel) { throw new Exception("Operation cancelled"); }

						var romConfigFile = Path.Join(romFolder, $"{game}.zip.cfg");

						// get rom config content
						if (string.IsNullOrEmpty(romConfigContent)) {
							if (data.overwrite || !File.Exists(romConfigFile)) {
								// file doesn't exist or we'll overwrite it
								romConfigContent = await downloaderService.DownloadFileText(pack.Repository, $"{pack.Roms.Src}/{game}.zip.cfg");

								// fix resolution and paths
								romConfigContent = ChangeResolution(romConfigContent, data.ratio);
								romConfigContent = FixPaths(romConfigContent, pack);
							}
							else {
								// file exist, we don"t overwrite: read it
								romConfigContent = await File.ReadAllTextAsync(romConfigFile);
							}
						}

						// write rom config
						if (data.overwrite || !File.Exists(romConfigFile)) {
							if (messageHandler.MustCancel) { throw new Exception("Operation cancelled"); }

							await File.WriteAllTextAsync(romConfigFile, romConfigContent);
							installed++;
						}
					}

					if (messageHandler.MustCancel) { throw new Exception("Operation cancelled"); }

					messageHandler.Progress($"{game}: download overlay (config)", total, current);

					// extract the overlay file name
					var overlayPath = GetCfgData(romConfigContent, "input_overlay");
					if (string.IsNullOrWhiteSpace(overlayPath)) { throw new FileNotFoundException($"Unable to parse rom config {game} to find overlay (input_overlay)"); }
					var overlayFi = new FileInfo(overlayPath);
					var overlayConfigDest = Path.Join(data.configFolder, overlayFi.Name);

					// download the overlay file name and extract the image file name
					var overlayConfigContent = string.Empty;
					if (data.overwrite || !File.Exists(overlayConfigDest)) {
						if (messageHandler.MustCancel) { throw new Exception("Operation cancelled"); }

						overlayConfigContent = await downloaderService.DownloadFileText(pack.Repository, $"{pack.Overlays.Src}/{overlayFi.Name}");

						// fix path
						overlayConfigContent = FixPaths(overlayConfigContent, pack);

						await File.WriteAllTextAsync(overlayConfigDest, overlayConfigContent);
					}
					else {
						overlayConfigContent = File.ReadAllText(overlayConfigDest);
					}

					if (messageHandler.MustCancel) { throw new Exception("Operation cancelled"); }

					messageHandler.Progress($"{game}: download overlay (image)", total, current);

					// extract the image file name
					var imagePath = GetCfgData(overlayConfigContent, "overlay0_overlay");
					if (string.IsNullOrWhiteSpace(imagePath)) { throw new FileNotFoundException($"Unable to parse overlay config {game} to find image (overlay0_overlay)"); }
					var imageFi = new FileInfo(imagePath);
					var imageDest = Path.Join(data.configFolder, imageFi.Name);

					// download the image
					if (data.overwrite || !File.Exists(imageDest)) {
						if (messageHandler.MustCancel) { throw new Exception("Operation cancelled"); }
						
						await downloaderService.DownloadFile(pack.Repository, $"{pack.Overlays.Src}/{imageFi.Name}", imageDest);
					}
				}

				messageHandler.Done($"Installed {installed} overlays", null);
			}
			catch (Exception ex) {
				messageHandler.Error(ex);
			}
		}

		/// <summary>
		/// Builds a regex string to get the specified key value
		/// </summary>
		/// <param name="key">The key.</param>
		/// <returns>The corresponding value</returns>
		private static string BuildCfgRegex(string key) {
			/// searched value looks like:
			/// key = "value"
			/// with or without spaces, with or without quotes, with or without trailing spaces
			return $"^{key}\\s*=\\s?\"?([^\"\\n]*)\"?\\s*$";
		}

		/// <summary>
		/// Changes the resolution in a file content.
		/// </summary>
		/// <param name="content">The file content.</param>
		/// <param name="ratio">The ratio.</param>
		/// <returns>The modified content</returns>
		private static string ChangeResolution(string content, float ratio) {
			if (ratio == 1) {
				return content;
			}

			var parameters = new List<string> {
				"custom_viewport_width",
				"custom_viewport_height",
				"custom_viewport_x",
				"custom_viewport_y",
				"video_fullscreen_x",
				"video_fullscreen_y"
			};

			foreach (var p in parameters) {
				var regex = new Regex(BuildCfgRegex(p), RegexOptions.Multiline | RegexOptions.IgnoreCase);
				var foundValue = regex.Match(content).Groups[1].Value;
				if (double.TryParse(foundValue, out var value)) {
					value = Math.Round(value * ratio, 0);
					content = regex.Replace(content, $"{p} = {(int)value}");
				}
			}

			return content;
		}

		/// <summary>
		/// Fixes the paths in a config.
		/// </summary>
		/// <param name="content">The config content.</param>
		/// <param name="pack">The pack to download.</param>
		/// <returns>
		/// The fixed content
		/// </returns>
		private static string FixPaths(string content, OverlayBundle pack) {
			if (ArcadeManagerEnvironment.SettingsOs != pack.BaseOs) {
				return content.Replace(pack.Base[pack.BaseOs], pack.Base[ArcadeManagerEnvironment.SettingsOs], StringComparison.InvariantCultureIgnoreCase);
			}

			return content;
		}

		/// <summary>
		/// Gets data from the specified config file.
		/// </summary>
		/// <param name="fileContent">The content of the file.</param>
		/// <param name="key">The key to look for.</param>
		/// <returns>
		/// The config value
		/// </returns>
		private static string GetCfgData(string fileContent, string key) {
			var match = Regex.Match(fileContent, BuildCfgRegex(key), RegexOptions.Multiline);
			if (match.Success && match.Captures.Any()) {
				return match.Groups[1].Value.Trim();
			}

			return null;
		}

		/// <summary>
		/// Gets the list of roms to process and their folder(s).
		/// </summary>
		/// <param name="romFolders">The rom folders.</param>
		/// <param name="entries">The available entries.</param>
		/// <returns>
		/// The roms to process
		/// </returns>
		private static IEnumerable<RomToProcess> GetRomsToProcess(string[] romFolders, IEnumerable<GithubTree.Entry> entries) {
			var result = new List<RomToProcess>();

			// list all the folders (arcade, fba, mame...)
			foreach (var folder in romFolders) {
				var di = new DirectoryInfo(folder);
				// get all rom files
				foreach (var fi in di.GetFiles("*.zip")) {
					var game = fi.Name.Replace(".zip", "");

					// only process files that are in the overlays pack
					if (entries.Any(e => e.Path.EndsWith($"{game}.zip.cfg", StringComparison.InvariantCultureIgnoreCase))) {
						var existing = result.Where(r => r.Game.Equals(game, StringComparison.InvariantCultureIgnoreCase)).FirstOrDefault();
						if (existing != null) {
							existing.TargetFolder.Add(fi.DirectoryName);
						}
						else {
							result.Add(new RomToProcess { Game = game, TargetFolder = new List<string> { fi.DirectoryName } });
						}
					}
				}
			}

			return result;
		}

		/// <summary>
		/// Downloads the common files.
		/// </summary>
		/// <param name="pack">The overlay pack to download.</param>
		/// <param name="destination">The destination path.</param>
		/// <param name="overwrite">if set to <c>true</c> overwrite existing files.</param>
		/// <param name="ratio">The ratio to change resolution.</param>
		/// <param name="messageHandler">The message handler.</param>
		/// <param name="total">The total number of items.</param>
		/// <param name="current">The current item number.</param>
		private async Task DownloadCommon(OverlayBundle pack, string destination, bool overwrite, float ratio, IMessageHandler messageHandler, int total, int current) {
			if (messageHandler.MustCancel) { return; }

			if (pack.Common != null && !string.IsNullOrEmpty(pack.Common.Src)) {
				IEnumerable<string> files = await downloaderService.DownloadFolder(pack.Repository, pack.Common.Src, destination, overwrite, (entry) => {
					messageHandler.Progress($"downloading {entry.Path}", total, current);
				});

				foreach (var f in files.Where(f => f.EndsWith(".cfg", StringComparison.InvariantCultureIgnoreCase))) {
					var fi = new FileInfo(f);
					messageHandler.Progress($"fixing {fi.Name}", total, current);

					if (overwrite || !fi.Exists) {
						var content = await File.ReadAllTextAsync(f);

						content = ChangeResolution(content, ratio);
						content = FixPaths(content, pack);

						await File.WriteAllTextAsync(f, content);
					}
				};
			}
		}

		/// <summary>
		/// A rom to process
		/// </summary>
		private class RomToProcess {

			/// <summary>
			/// Gets or sets the name of the game.
			/// </summary>
			public string Game { get; set; }

			/// <summary>
			/// Gets or sets the folders in which the game is present.
			/// </summary>
			public List<string> TargetFolder { get; set; } = new();
		}
	}
}
